sdks/java/core/src/test/java/org/apache/beam/sdk/metrics/MetricsTest.java

```diff
@@ -69,216 +74,286 @@ private static MetricQueryResults queryTestMetrics(PipelineResult result) {
+    public void testDistributionToCell() {
+      MetricsContainer mockContainer = Mockito.mock(MetricsContainer.class);
+      Distribution mockDistribution = Mockito.mock(Distribution.class);
+      when(mockContainer.getDistribution(METRIC_NAME)).thenReturn(mockDistribution);
+
+      Distribution distribution = Metrics.distribution(NS, NAME);
+
+      MetricsEnvironment.setCurrentContainer(mockContainer);
+      distribution.update(5L);
+
+      verify(mockDistribution).update(5L);
+
+      distribution.update(36L);
+      distribution.update(1L);
+      verify(mockDistribution).update(36L);
+      verify(mockDistribution).update(1L);
+    }
+
+    @Test
+    public void testCounterToCell() {
+      MetricsContainer mockContainer = Mockito.mock(MetricsContainer.class);
+      Counter mockCounter = Mockito.mock(Counter.class);
+      when(mockContainer.getCounter(METRIC_NAME)).thenReturn(mockCounter);
+
+      Counter counter = Metrics.counter(NS, NAME);
+
+      MetricsEnvironment.setCurrentContainer(mockContainer);
+      counter.inc();
+      verify(mockCounter).inc(1);
+
+      counter.inc(47L);
+      verify(mockCounter).inc(47);
+
+      counter.dec(5L);
+      verify(mockCounter).inc(-5);
+    }
   }
 
-  @Test
-  public void testDistributionWithEmptyNamespace() {
-    thrown.expect(IllegalArgumentException.class);
-    Metrics.distribution("", NAME);
+  /** Tests for committed metrics. */
+  @RunWith(JUnit4.class)
+  public static class CommittedMetricTests extends SharedTestBase {
+    @Category({ValidatesRunner.class, UsesCommittedMetrics.class, UsesCounterMetrics.class,
+        UsesDistributionMetrics.class, UsesGaugeMetrics.class})
+    @Test
+    public void testAllCommittedMetrics() {
+      PipelineResult result = runPipelineWithMetrics();
+      MetricQueryResults metrics = queryTestMetrics(result);
+
+      assertAllMetrics(metrics, true);
+    }
+
+    @Category({ValidatesRunner.class, UsesCommittedMetrics.class, UsesCounterMetrics.class})
+    @Test
+    public void testCommittedCounterMetrics() {
+      PipelineResult result = runPipelineWithMetrics();
+      MetricQueryResults metrics = queryTestMetrics(result);
+      assertCounterMetrics(metrics, true);
+    }
+
+    @Category({ValidatesRunner.class, UsesCommittedMetrics.class, UsesDistributionMetrics.class})
+    @Test
+    public void testCommittedDistributionMetrics() {
+      PipelineResult result = runPipelineWithMetrics();
+      MetricQueryResults metrics = queryTestMetrics(result);
+      assertDistributionMetrics(metrics, true);
+    }
+
+    @Category({ValidatesRunner.class, UsesCommittedMetrics.class, UsesGaugeMetrics.class})
+    @Test
+    public void testCommittedGaugeMetrics() {
+      PipelineResult result = runPipelineWithMetrics();
+      MetricQueryResults metrics = queryTestMetrics(result);
+      assertGaugeMetrics(metrics, true);
+    }
+
+    @Test
+    @Category({NeedsRunner.class, UsesAttemptedMetrics.class, UsesCounterMetrics.class})
```

Comment:
I think this one is a `ValidatesRunner` because `PipelineResults.metrics()` actually uses `MetricsContainerStepMap` which is a runner-core componant and which is set by the different runners with their own aggregators to support metrics.

Reply:
For `MetricsTest`, I tried to reduce what I felt was redundant validation of the runner-provided pieces. I kept a few `@ValidatesRunner` tests exercise different parts of the `MetricsContainerStepMap`; in particular:

* `testAllAttemptedMetrics()`
* `testAllCommittedMetrics()`
* `testAttempted[Counter|Distribution|Gauge]Metrics()`
* `testCommitted[Counter|Distribution|Gauge]Metrics()`

This test, `testBoundedSourceMetrics()` is validating that metrics from `BoundedSources` get added into the container. I don't believe this exercises any new runner-behavior not already covered by the other tests. Which is why I converted it.

Let me know if you disagree. If there is runner-behavior that needs to be validated we shouldn't be shy about keeping it as `@ValidatesRunner`. I think it's better to have long test runs than gaps in our validation.

Reply:
A bit late for this comment but +1:
indeed the tests you pointed actually test the containers and you left them annotated with ValidatesRunner, so it is ok. You're right to say that the other tests that use `PipelineResults.metrics()` test other parts that the metrics aggregation with the runners.
Also I agree to keep validatesRunner tests as short as possible and remove any redundancy; only one validatesRunner test is enough to test the metrics aggregation.